cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,415 @@
+
+    const FString& type = joint_node->GetAttribute(TEXT("type"));
+    if (type.Equals(TEXT("revolute"))) {
+        joint_desc.type_ = UrdfJointType::Revolute;
+    } else if (type.Equals(TEXT("continuous"))) {
+        joint_desc.type_ = UrdfJointType::Continuous;
+    } else if (type.Equals(TEXT("prismatic"))) {
+        joint_desc.type_ = UrdfJointType::Prismatic;
+    } else if (type.Equals(TEXT("fixed"))) {
+        joint_desc.type_ = UrdfJointType::Fixed;
+    } else if (type.Equals(TEXT("floating"))) {
+        joint_desc.type_ = UrdfJointType::Floating;
+    } else if (type.Equals(TEXT("planar"))) {
+        joint_desc.type_ = UrdfJointType::Planar;
+    } else {
+        ASSERT(false);
+    }
+
+    bool has_origin_node = false;
+    bool has_parent_node = false;
+    bool has_child_node = false;
+    bool has_axis_node = false;
+    bool has_calibration_node = false;
+    bool has_dynamics_node = false;
+    bool has_limit_node = false;
+    bool has_mimic_node = false;
+    bool has_safety_controller_node = false;
+
+    for (auto& node : joint_node->GetChildrenNodes()) {
+        const FString& tag = node->GetTag();
+        if (tag.Equals(TEXT("origin"))) {
+            ASSERT(!has_origin_node);
+            has_origin_node = true;
+            joint_desc.origin_.SetLocation(parseVector(node->GetAttribute("xyz")));
+            joint_desc.origin_.SetRotation(parseRotation(node->GetAttribute("rpy")));
+        } else if (tag.Equals(TEXT("parent"))) {
+            ASSERT(!has_parent_node);
+            has_parent_node = true;
+            joint_desc.parent_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("link")));
+        } else if (tag.Equals(TEXT("child"))) {
+            ASSERT(!has_child_node);
+            has_child_node = true;
+            joint_desc.child_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("link")));
+        } else if (tag.Equals(TEXT("axis"))) {
+            ASSERT(!has_axis_node);
+            has_axis_node = true;
+            joint_desc.axis_ = parseVector(node->GetAttribute(TEXT("xyz")));
+        } else if (tag.Equals(TEXT("calibration"))) {
+            ASSERT(!has_calibration_node);
+            has_calibration_node = true;
+            for (auto& attr_node : node->GetAttributes()) {
+                if (attr_node.GetTag() == "rising") {
+                    ASSERT(joint_desc.calibration_type_ == CalibrationType::Invalid);
+                    joint_desc.calibration_type_ = CalibrationType::Rising;
+                    joint_desc.rising_ = FCString::Atof(*attr_node.GetValue());
+                } else if (attr_node.GetTag() == "falling") {
+                    ASSERT(joint_desc.calibration_type_ == CalibrationType::Invalid);
+                    joint_desc.calibration_type_ = CalibrationType::Falling;
+                    joint_desc.falling_ = FCString::Atof(*attr_node.GetValue());
+                }
+            }
+        } else if (tag.Equals(TEXT("dynamics"))) {
```

Comment:
- `damping` and `friction` are listed as optional in the spec. Will this code work as expected if they're omitted?

Reply:
if attr do not exist, GetAttr will return an empty string, and `FCString::Atof` return zero. Parameters with default value of zero is fine, and check existance is needed for others. Or we could have a special function to handle this situation.